runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,789 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.puzzlesolver.solver.pathfinding;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import net.runelite.api.Point;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.BLANK_TILE_VALUE;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.DIMENSION;
+import net.runelite.client.plugins.puzzlesolver.solver.PuzzleState;
+import net.runelite.client.plugins.puzzlesolver.solver.heuristics.Heuristic;
+
+public class IDAStarMM extends IDAStar
+{
+	private PuzzleState currentState;
+	private List<PuzzleState> stateList = new ArrayList<>();
+	private List<List<Integer>> validRowNumbers = new ArrayList<>();
+	private List<List<Integer>> validColumnNumbers = new ArrayList<>();
+
+	public IDAStarMM(Heuristic heuristic)
+	{
+		super(heuristic);
+
+		//Add valid numbers for rows and columns
+		validRowNumbers.add(Arrays.asList(0, 1, 2, 3, 4));
+		validRowNumbers.add(Arrays.asList(6, 7, 8, 9));
+		validColumnNumbers.add(Arrays.asList(5, 10, 15, 20));
+	}
+
+	@Override
+	public List<PuzzleState> computePath(PuzzleState root)
+	{
+		currentState = root;
+		stateList.add(root);
+
+		List<PuzzleState> path = new ArrayList<>();
+
+		//Reduce to 4x5
+		solveRow(0);
+
+		//Reduce to 4x4
+		solveColumn();
+
+		//Reduce to 3x4
+		solveRow(1);
+
+		//Remove last state
+		stateList.remove(stateList.size() - 1);
+
+		//Pathfinder for 4x4
+		path.addAll(super.computePath(currentState));
+
+		path.addAll(0, stateList);
+
+		return path;
+	}
+
+	private void solveRow(int row)
+	{
+		for (int i = row; i < DIMENSION; i++)
+		{
+			int valTarget = row * DIMENSION + i;
+
+			int valCurrent = currentState.getPiece(i, row);
+
+			if (valCurrent != valTarget)
+			{
+				Point locTarget = new Point(i, row);
+				moveTowardsVal(valTarget, locTarget,  true);
```

Comment:
I don't think there is any reason to use Point all over.

Consider just passing in ``valTarget``, ``i`` and ``row`` and remove the Point usage.